ocaml/xapi/test_network_event_loop.ml

```diff
@@ -0,0 +1,165 @@
+let test_network_event_loop ~no_nbd_networks_at_start () =
+  let __context, _ = Test_event_common.event_setup_common () in
+  (* We need to set test_mode to true to ensure that the event loop will use
+     the local Xapi_events.from function instead of the Client module, which
+     would cause the test to fail. *)
+  Helpers.test_mode := true;
+
+  let localhost = Helpers.get_localhost ~__context in
+  let other_host = Test_common.make_host ~__context () in
+
+  (* We have to wait for a bit for the event loop to notice the changes, without a delay the test will fail. *)
+  let delay = 0.2 in
+  let network_event_loop_wait_after_failure_seconds = 0.4 in
+  let received_params = ref None in
+
+  (* We simulate failure of the firewall update script this way *)
+  let fail_firewall_update = ref false in
+
+  let start_event_loop = Thread.create
+      (fun () ->
+         Network_event_loop._watch_networks
+           __context
+           ~update_firewall:(fun pifs ->
+               if !fail_firewall_update then failwith "Failed to update firewall";
+               received_params := Some pifs
+             )
+           ~wait_after_failure_seconds:network_event_loop_wait_after_failure_seconds
+      )
+  in
+
```

Comment:
There's a common pattern of
```
fail_firewall_update := None;
...do setup stuff...
assert_received_params msg expected;
```
(or `assert_not_called` at the end) so I wonder about changing those two functions to be like this:
```
let assert_received_params setup msg expected =
  received_params := None;
  setup ();
  Thread.delay delay;
  ...examine received_params with OUnit.assert_something...
in
```

Reply:
I agree with you that the test is not ideal right now, and I'll improve it once I can think of a better way :disappointed: I wanted to test the long-running behaviour of this event loop, and also how it behaves in case of failures.